flytecopilot/data/download.go

```diff
@@ -31,24 +33,94 @@ type Downloader struct {
 	mode core.IOStrategy_DownloadMode
 }
 
-// TODO add support for multipart blobs
-func (d Downloader) handleBlob(ctx context.Context, blob *core.Blob, toFilePath string) (interface{}, error) {
+func (d Downloader) handleBlob(ctx context.Context, blob *core.Blob, toPath string) (interface{}, error) {
 	ref := storage.DataReference(blob.Uri)
 	scheme, _, _, err := ref.Split()
 	if err != nil {
 		return nil, errors.Wrapf(err, "Blob uri incorrectly formatted")
 	}
+
+	if blob.GetMetadata().GetType().Dimensionality == core.BlobType_MULTIPART {
+		items, err := d.store.GetItems(ctx, ref)
+		if err != nil || len(items) == 0 {
+			logger.Errorf(ctx, "failed to collect items from multipart blob [%s]", ref)
+			return nil, err
+		}
+
+		success := 0
+		var mu sync.Mutex
+		var wg sync.WaitGroup
+		for _, absPath := range items {
+			// capture range variable
+			absPath := absPath
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				defer func() {
+					if err := recover(); err != nil {
+						logger.Errorf(ctx, "recover receives error: %s", err)
+					}
+				}()
+
+				ref := storage.DataReference(absPath)
+				reader, err := DownloadFileFromStorage(ctx, ref, d.store)
+				if err != nil {
+					logger.Errorf(ctx, "Failed to download from ref [%s]", ref)
+					return
+				}
+				defer func() {
+					err := reader.Close()
+					if err != nil {
+						logger.Errorf(ctx, "failed to close Blob read stream @ref [%s]. Error: %s", ref, err)
+					}
+				}()
+
+				_, _, relativePath, err := ref.Split()
+				if err != nil {
+					logger.Errorf(ctx, "Failed to parse ref [%s]", ref)
+					return
+				}
+				newPath := filepath.Join(toPath, relativePath)
+				dir := filepath.Dir(newPath)
+
+				mu.Lock()
+				// 0755: the directory can be read by anyone but can only be written by the owner
+				os.MkdirAll(dir, 0755)
+				writer, err := os.Create(newPath)
+				mu.Unlock()
+				if err != nil {
```

Comment:
technically this line can be moved up one right?  the lock is only for folder creation?  or is it needed for file creation also?

Reply:
Yes! There's no need to protect `os.Create` in this case. Will correct this.